### ISSUE TITLE
CKR_MECHANISM_INVALID error for 3DES wrap/unwrap

### ISSUE THREAD

**Issue by justincranford** [opendnssec/SoftHSMv2#596]
Please add missing 3DES wrap/unwrap support to SoftHSM2.

I ran into an issuing using SoftHSM2 to test PKIX-CMP user enroll. During enroll, SoftHSM2 is asked to do a 3DES unwrap which fails due to CKR_MECHANISM_INVALID.

I used OpenSC pkcs11-tool.exe to list supported mechanisms in SoftHSM2 (versus a physical HSM such as nCipher):
- SoftHSM2: DES3-ECB, encrypt, decrypt
- nCipher: DES3-ECB, keySize={24,24}, hw, encrypt, decrypt, wrap, unwrap, derive

It seems that SoftHSM2 is missing 4 out of the 6 3DES operations: wrap, unwrap, derive, and hw.

**Comment by rijswijk** [comment]
SoftHSM only supports symmetric key-based unwrapping with AES, and not with 3DES. There are no plans to implement this, as 3DES is a legacy algorithm that should not be used anymore. I would strongly recommend that you use AES in your application instead.

If you really want to support 3DES unwrapping, it should not be hard to implement, as you can essentially copy most of the code from the AES-based codebase.

**Comment by justincranford** [comment]
I agree AES wrap/unwrap is preferred. Newer CA versions can and should use AES wrap. The issue is older CAs have a way of sticking around for a long time due to 10-year and 20-year CA certs. Older CAs are stuck using 3DES wrap/unwrap.

It was a long shot asking for 3DES wrap/unwrap. It would be nice to have for regression testing small fixes for older PKIX-CMP CAs.

Thank you.

**Comment by hanhsuan** [comment]
I have quickly implemented 3DES wrap/unwrap in [my fork](https://github.com/hanhsuan/SoftHSMv2/tree/fix_issue_596), and test with a few conditions that I need only. If someone could help me to test it will be great.
